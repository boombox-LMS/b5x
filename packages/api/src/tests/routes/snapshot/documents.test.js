const supertest = require("supertest");
const { z } = require("zod");
const { DocumentContentsSchema, UserResponseSchema } = require("@b5x/types");

describe("Documents routes should return a 200", () => {
  let app;
  let cookie;
  let testDocumentUri;
  const apiPrefix = global.__SUPERTEST_API_PREFIX__;

  beforeAll(async () => {
    // Add a seeded test database to the app, available as app.get('db')
    app = await global.buildAppWithSeededTestDatabase({
      users: [{ persona: "completed" }],
    });

    // TODO: Have seeding functionality in the parser to keep all of these URIs consistent
    // even as the demo content is updated
    testDocumentUri = "smoke-testing-vseed_setup";

    // Log the user in (someone who has completed all topics)
    await supertest(app)
      .post(apiPrefix + "users.logIn")
      .send({ email: "completed-1@test.com" })
      .then((res) => {
        // Save the cookie to send with later requests
        cookie = global.getCookie(res) || cookie;
      });
  });

  // Close the database connection, which allows Jest to exit gracefully
  afterAll(async () => {
    await app.get("db").destroy();
  });

  // documents.contents ---------------------------------------------

  describe("documents.contents matches expectations", () => {
    let responseBody;

    test("documents.contents returns a 200", async () => {
      await supertest(app)
        .get(apiPrefix + `documents.contents?documentUri=${testDocumentUri}`)
        .set("Cookie", cookie)
        .expect(200)
        .then((res) => {
          responseBody = res.body;
        });
    });

    test("documents.contents returns the correct data type", () => {
      const validator = () => {
        DocumentContentsSchema.parse(responseBody);
      };
      expect(validator).not.toThrowError();
    });

    test("documents.contents matches the snapshot on file", () => {
      expect(responseBody).toMatchSnapshot();
    });
  });

  // documents.responses --------------------------------------------

  describe("documents.responses matches expectations", () => {
    let responseBody;

    test("documents.responses returns a 200", async () => {
      await supertest(app)
        .get(apiPrefix + `documents.responses?documentUri=${testDocumentUri}`)
        .set("Cookie", cookie)
        .expect(200)
        .then((res) => {
          responseBody = res.body;
        });
    });

    test("documents.responses returns the correct data type", () => {
      // TODO: Not sure why this fails when a minimum length string key is defined
      const ResponseBodySchema = z.record(UserResponseSchema);
      const validator = () => {
        ResponseBodySchema.parse(responseBody);
      };

      expect(validator).not.toThrowError();
    });

    test("documents.responses matches the snapshot on file", () => {
      expect(responseBody).toMatchSnapshot();
    });
  });

  // documents.verifyCompletion -------------------------------------

  describe("documents.verifyCompletion matches the snapshot on file", () => {
    let responseBody;

    test("documents.verifyCompletion returns a 200", async () => {
      await supertest(app)
        .post(apiPrefix + `documents.verifyCompletion`)
        .send({ documentUri: testDocumentUri })
        .set("Cookie", cookie)
        .expect(200)
        .then((res) => {
          responseBody = res.body;
        });
    });

    test("documents.verifyCompletion returns the correct data type", () => {
      const ResponseBodySchema = z.object({
        documentUri: z.string().min(1),
        documentIsCompleted: z.boolean(),
      });
      const validator = () => {
        ResponseBodySchema.parse(responseBody);
      };

      expect(validator).not.toThrowError();
    });

    test("documents.verifyCompletion matches the snapshot on file", () => {
      expect(responseBody).toMatchSnapshot();
    });
  });
});
