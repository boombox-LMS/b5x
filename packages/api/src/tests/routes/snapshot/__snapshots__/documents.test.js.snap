// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documents routes should return a 200 documents.contents matches expectations documents.contents matches the snapshot on file 1`] = `
{
  "contentMode": "tutorial",
  "fragments": [
    {
      "children": [],
      "contentType": "HtmlContent",
      "contents": "<h1 id="setup">Setup</h1>
<p>Please seed the database before proceeding further, just to verify that you&#39;re working with fresh data.</p>
",
      "data": {
        "isTemplated": false,
      },
      "dependencyUris": [],
      "displayConditions": [],
      "documentUri": "smoke-testing-vseed_setup",
      "isRequired": false,
      "isStateful": false,
      "uri": "smoke-testing-vseed_setup_f1",
    },
    {
      "children": [
        {
          "children": [],
          "contentType": "HtmlContent",
          "contents": "<p>The database is freshly seeded.</p>
",
          "data": {
            "isTemplated": false,
          },
          "dependencyUris": [],
          "displayConditions": [],
          "documentUri": "smoke-testing-vseed_setup",
          "isRequired": false,
          "isStateful": false,
          "uri": "smoke-testing-vseed_setup_f3",
        },
      ],
      "contentType": "ContinueButton",
      "contents": "",
      "data": {},
      "dependencyUris": [],
      "displayConditions": [],
      "documentUri": "smoke-testing-vseed_setup",
      "isRequired": true,
      "isStateful": true,
      "uri": "smoke-testing-vseed_setup_f2",
    },
    {
      "children": [],
      "contentType": "DocumentCompletionContent",
      "contents": "",
      "data": {},
      "dependencyUris": [
        "smoke-testing-vseed_setup_f2",
      ],
      "displayConditions": [
        {
          "data": {
            "operator": "status-is",
            "targetValue": "completed",
            "uri": "smoke-testing-vseed_setup_f2",
          },
          "type": "condition",
        },
      ],
      "documentUri": "smoke-testing-vseed_setup",
      "isRequired": false,
      "isStateful": false,
      "uri": "smoke-testing-vseed_setup_f4",
    },
  ],
  "topicUri": "smoke-testing-vseed",
  "uri": "smoke-testing-vseed_setup",
}
`;

exports[`Documents routes should return a 200 documents.responses matches expectations documents.responses matches the snapshot on file 1`] = `
{
  "smoke-testing-vseed_setup_f2": {
    "fragmentUri": "smoke-testing-vseed_setup_f2",
    "status": "completed",
    "value": "t",
  },
}
`;

exports[`Documents routes should return a 200 documents.verifyCompletion matches the snapshot on file documents.verifyCompletion matches the snapshot on file 1`] = `
{
  "documentIsCompleted": true,
  "documentUri": "smoke-testing-vseed_setup",
}
`;
